api/base/serializers.py

```diff
@@ -144,7 +189,86 @@ def to_representation(self, value):
             else:
                 meta[key] = _rapply(self.meta[key], _url_val, obj=value, serializer=self.parent)
 
-        return {'links': {self.link_type: {'href': url, 'meta': meta}}}
+        return self.format_json_response(link_type=self.link_type, url=url, meta=meta)
+
+
+class JSONAPIHyperlinkedRelatedField(ser.HyperlinkedRelatedField, HyperlinkedFieldMixin):
+    """
+    HyperlinkedRelated field that returns a nested dict with url,
+    optional meta information, and link_type.
+
+    Example:
+
+        branched_from = JSONAPIHyperlinkedRelatedField(view_name='nodes:node-detail', lookup_field='pk',
+                                                    lookup_url_kwarg='node_id', read_only=True, link_type='related')
+
+    """
+    def __init__(self, view_name=None, **kwargs):
+        self.meta = kwargs.pop('meta', {})
+        self.link_type = kwargs.pop('link_type', 'url')
+        super(JSONAPIHyperlinkedRelatedField, self).__init__(view_name=view_name, **kwargs)
+
+    def to_representation(self, value):
+        """
+        Returns nested dictionary in format {'links': {'self.link_type': ... }
+
+        If no meta information, self.link_type is equal to a string containing link's URL.  Otherwise,
+        the link is represented as a links object with 'href' and 'meta' members.
+        """
+        url = super(JSONAPIHyperlinkedRelatedField, self).to_representation(value)
+        meta = _rapply(self.meta, _url_val, obj=value, serializer=self.parent)
+        return self.format_json_response(link_type=self.link_type, url=url, meta=meta)
+
+
+class JSONAPIHyperlinkedGuidRelatedField(ser.Field, HyperlinkedFieldMixin):
+    """
+    Field that returns a nested dict with the url (constructed based
+    on the object's type), optional meta information, and link_type.
+
+    Example:
+
+        target = JSONAPIHyperlinkedGuidRelatedField(link_type='related', meta={'type': 'get_target_type'})
+
+    """
+    def __init__(self, **kwargs):
+        self.meta = kwargs.pop('meta', {})
+        self.link_type = kwargs.pop('link_type', 'url')
+        super(JSONAPIHyperlinkedGuidRelatedField, self).__init__(read_only=True, **kwargs)
+
+    def get_guid_views(self, guid):
+        # get target type from guid
+        guid_object = Guid.load(guid).referent
+        if not guid_object:
+            raise NotFound
+
+        if isinstance(guid_object, Node):
+            self.link_type = 'related'
+            guid_views = GUID_VIEWS['node']
+
+        elif isinstance(guid_object, Comment):
+            self.link_type = 'self'
+            guid_views = GUID_VIEWS['comment']
+
+        else:
+            raise InvalidModelValueError('Invalid comment target.')
```

Comment:
Just for reference: If this is so unusual that it should never ever ever happen, an `assert` statement is also a valid strategy.